cmd/csi-snapshotter/main.go

```diff
@@ -1,7 +1,217 @@
+	kubeconfig                   = flag.String("kubeconfig", "", "Absolute path to the kubeconfig file. Required only when running out of cluster.")
+	resync                       = flag.Duration("resync", 10*time.Second, "Resync interval of the controller.")
+	connectionTimeout            = flag.Duration("connection-timeout", 1*time.Minute, "Timeout for waiting for CSI driver socket.")
+	csiAddress                   = flag.String("csi-address", "/run/csi/socket", "Address of the CSI driver socket.")
+	createSnapshotDataRetryCount = flag.Int("createSnapshotDataRetryCount", 5, "Number of retries when we create a snapshot data object for a snapshot.")
+	createSnapshotDataInterval   = flag.Duration("createSnapshotDataInterval", 10*time.Second, "Interval between retries when we create a snapshot data object for a snapshot.")
+	resyncPeriod                 = flag.Duration("resyncPeriod", 60*time.Second, "The period that should be used to re-sync the snapshot.")
+)
 
 func main() {
-	fmt.Println("vim-go")
+	flag.Set("logtostderr", "true")
+	flag.Parse()
+
+	vs := &crdv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:         "kk",
+			Name:              "test",
+			SelfLink:          "/apis/volumesnapshot.csi.k8s.io/v1alpha1/namespaces/default/volumesnapshots/snapshot-demo",
+			UID:               "046722da-6084-11e8-aa97-fa163ec505d6",
+			ResourceVersion:   "514",
+			CreationTimestamp: metav1.Now(),
+		},
+		Spec: crdv1.VolumeSnapshotSpec{
+			PersistentVolumeClaimName: "testpvc",
+			SnapshotClassName:         "testsnapclass",
+		},
+	}
+
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(vs)
+	if err != nil {
+		glog.Error(err.Error())
+	}
+
+	glog.Info(key)
+	glog.Info(vs)
+
+	// Create the client config. Use kubeconfig if given, otherwise assume in-cluster.
+	config, err := buildConfig(*kubeconfig)
+	if err != nil {
+		glog.Error(err.Error())
+		os.Exit(1)
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		glog.Error(err.Error())
+		os.Exit(1)
+	}
+
+	snapClient, err := clientset.NewForConfig(config)
+	if err != nil {
+		glog.Errorf("Error building snapshot clientset: %s", err.Error())
+		os.Exit(1)
+	}
+
+	factory := informers.NewSharedInformerFactory(snapClient, *resync)
+
+	// Create CRD resource
+	aeclientset, err := apiextensionsclient.NewForConfig(config)
+	if err != nil {
+		panic(err)
+	}
+
+	// initialize CRD resource if it does not exist
+	err = CreateCRD(aeclientset)
+	if err != nil {
+		panic(err)
+	}
+
+	// make a new config for our extension's API group, using the first config as a baseline
+	crdClient, _, err := NewClient(config)
+	if err != nil {
+		panic(err)
+	}
+
+	// wait until CRD gets processed
+	err = WaitForSnapshotResource(crdClient)
+	if err != nil {
+		panic(err)
```

Comment:
why use panic instead of os.Exist like below?

Reply:
Will change it.